[flang][NFC] speedup BoxedProcedure for derived types with many components

jeanPerier opened the pull request:
This patch speeds up the compilation time of the example in https://github.com/llvm/llvm-project/issues/76478#issuecomment-2011023289 from 2 minutes with my builds to about 2 seconds.

MLIR timers showed more than 98% of the time was spend in BoxedProcedure trying to figure out if a type needs to be converted.

This is because walking the fir.type members is very expansive for types containing many components and/or components with many sub-components.

Increase the caching time of visited types from "the type being visited" to "the whole pass". Use DenseMap since it is not ok anymore to assume this container will only have a few elements.

jeanPerier committed:
[flang][NFC] speedup BoxedProcedure for derived types with many components

This patch speeds up the compilation time of the example in
https://github.com/llvm/llvm-project/issues/76478#issuecomment-2011023289
from 2mins with my builds to about 2 seconds.

MLIR timers showed more than 98% of the time was spend in BoxedProcedure
trying to figure out if a type needs to be converted.

This is because walking the fir.type members is very expansive for types
containing many components and/or components with many sub-components.

Increase the caching time of visited types from "the type being visited"
to "the whole pass". Use DenseMap since it is not ok anymore to assume
this container will only have a few elements.

clementval reviewed (approved):
LGTM. Nice improvement!

ceseo reviewed (approved):
LGTM. This solves the case for both reproducers in #76478

yus3710-fj reviewed (approved):
LGTM. Thank you for the quick fix!

jeanPerier committed:
fix regression with indirect type recursion

The previous update caused t002 type in the added test to not
be rewritten because it was first analyzed when analyzing t001.
Since the reason it needs to be rewritten is because it contains
a t001, but t001 is not fully analyzed when first analyzing t002,
the first analyzed yieled false. This "wrong" result is not an
issue when analyzing t001 (since t002 does not determine t001
result). But caching this invalid result for later usage is invalid.

Add a boolean to only keep analysis for top-level types where the result
is known to be correct.
This restriction has no measurable impact on the speed-up: I still measure
2 seconds.
